[Mobile] Fix crash on undefined image url

mkevins opened the pull request:
## Description
<!-- Please describe what you have changed or added -->
This PR adds a check to ensure `attributes.url` exists before attempting to call `indexOf`. Currently, there are scenarios which can result in `attributes.id` being truthy, while `attributes.url` is `null` or `undefined` for an image block.

When an image is cancelled before it is finished uploading, the `url` is set to `null`, but the `id` of the block remains. The next time the post is edited, the local draft will contain the state which triggers the crash.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->
This was tested using the steps here: https://github.com/wordpress-mobile/WordPress-Android/issues/9768#issuecomment-490852531

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->
This is a bug fix for the crash discovered here: https://github.com/wordpress-mobile/WordPress-Android/issues/9768

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://wordpress.org/gutenberg/handbook/designers-developers/ -->

mkevins committed:
Add safety check to image url before calling indexOf

hypest commented:
> `attributes.id` being truthy, while `attributes.url` is null or undefined

👋 @mkevins , I wonder, is that combination a valid state? If not, maybe we also need to try to set the block to a valid state?

mkevins committed:
Set attributes.id to null when an image upload is cancelled

mkevins committed:
Add console.warn and comments to help track WordPress-Android issue #9768

mkevins commented:
> ￼ @mkevins , I wonder, is that combination a valid state? If not, maybe we also need to try to set the block to a valid state?

I don't think this state is "valid". I've added some commits to set `attributes.id` to `null` when an image upload is cancelled. Also, I've added a `console.warn` which may be useful in tracking, as I expect we will still encounter this state from anyone that has already experienced this crash. This is because the local draft will contain the state. This is why we must also keep the check to prevent the crash in those cases.

daniloercoli commented:
I've been chatting with @mkevins and helped a bit testing and finding other flows that can result in attributes.id being set, while attributes.url is null, and didn't find another flow yet. 
I think the current solution if OK and we can include this to an hotfix if we're planning to release it.

marecar3 commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -81,7 +81,14 @@ class ImageEdit extends React.Component {
 		const { attributes, setAttributes } = this.props;
 
-		if ( attributes.id && ! isURL( attributes.url ) ) {
+    // This will warn when we have `id` defined, while `url` is undefined.
```
This whole block of code if shifted left, please can you make sure that it's properly aligned?

Probably the best option to prevent those kinds of commits is to cd into `gutenberg-mobile/gutenberg` and run `npm install` - which will add hooks so after that whenever you want to commit something hooks will trigger lint checks and you won't be able to commit without resolving lint errors.

mkevins replied:
Thank you @marecar3 . I have pushed a fix for the lint errors. I added those lines from the "inner-inner" directory while working / testing with WPAndroid, so I didn't catch that I had spaces instead of tabs :grimacing: . Thanks for reviewing! :)

marecar3 reviewed (approved):
Hey @mkevins, I have left some comment but besides that, it's great to work here!
So be free to merge this PR when you resolve the comment.

mkevins committed:
Fix lint errors

mkevins commented:
Thank you @daniloercoli and @marecar3 for testing and reviewing!

koke commented:
> I don't think this state is "valid"

I was looking at [the code](https://github.com/wordpress-mobile/gutenberg-mobile/blob/develop/ios/gutenberg/MediaUploadCoordinator.swift#L30) and it seems intentional and valid to me. The `mediaID` is a local UUID-like identifier. If the upload fails or was canceled, the iOS app will still have a local media object representing the image, and can use that to render the image, even after a relaunch. I think the `url` would only be present after a successful upload?

This is just from a static analysis though, I haven't debugged this to validate, but I imagine @SergioEstevao will be able to confirm or deny this 😄

marecar3 commented:
Hey @koke I think that @mkevins solution is good. When a user clicks on the `Cancel upload` button, it will reset the image block to the default state ( media upload options are presented). But the problem is that in the code we have left `id` which is hanging there and which needs to be removed.

Also, a failed state is different then cancel state as it can retry uploading.

Btw, I found some bug on iOS regarding failed upload

https://github.com/wordpress-mobile/gutenberg-mobile/issues/978

cc : @SergioEstevao

koke commented:
> When a user clicks on the Cancel upload button, it will reset the image block to the default state ( media upload options are presented)

Considering the changes in #966, it’s not clear to me whether canceling would automatically bring the state back to placeholder, or allow to retry. But a failed upload would have a failed state where having the media ID still makes sense.

We could eventually decide something different, but that is why I think this is considered a valid state on the iOS code.

marecar3 commented:
> > When a user clicks on the Cancel upload button, it will reset the image block to the default state ( media upload options are presented)
> 
> Considering the changes in #966, it’s not clear to me whether canceling would automatically bring the state back to placeholder, or allow to retry. But a failed upload would have a failed state where having the media ID still makes sense.
> 
> We could eventually decide something different, but that is why I think this is considered a valid state on the iOS code.

Not sure if the link is good, but I agree with your thoughts :)

koke commented:
Ah, wrong repo 😂

https://github.com/wordpress-mobile/gutenberg-mobile/issues/966
